python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+            dataRef.dataId['visit'] = v2
+            exp2 = self.isr.runDataRef(dataRef).exposure
+            del dataRef.dataId['visit']
+            self._checkExpLengthEqual(exp1, exp2, v1, v2)
+
+            self.log.info('Preparing images for cross-correlation calculation for detector %s' % detNum)
+            # note the shape of these returns depends on level
+            _scaledMaskedIm1, _means1 = self._prepareImage(exp1, gains, self.config.level)
+            _scaledMaskedIm2, _means2 = self._prepareImage(exp2, gains, self.config.level)
+
+            if self.debug.enabled:
+                try:
+                    frameId1 = exp1.getMetadata().get("FRAMEID")
+                    frameId2 = exp2.getMetadata().get("FRAMEID")
+                    frameId = '_diff_'.join(frameId1, frameId2)
+                except:
+                    frameId = 'Im1 diff Im2'
+
+            # Compute the cross-correlation and means
+            # at the appropriate config.level:
+            # - "DETECTOR": one key, so compare the two visits to each other
+            # - "AMP": n_amp keys, comparing each amplifier of one visit
+            #          to the same amplifier in the visit its paired with
+            for det_object in _scaledMaskedIm1.keys():
+                xcorrs[det_object].append(self._crossCorrelate(_scaledMaskedIm1[det_object],
+                                                               _scaledMaskedIm2[det_object]))
+                means[det_object].append([_means1[det_object], _means2[det_object]])
+
+                # TODO: DM-15305 improve debug functionality here.
+                # This is position 1 for the removed code.
+
+        # generate the kernel(s)
+        self.log.info('Generating kernel(s) for %s' % detNum)
+        for det_object in xcorrs.keys():  # looping over either detectors or amps
+            if self.config.level == 'DETECTOR':
+                objId = 'detector %s' % det_object
+            elif self.config.level == 'AMP':
+                objId = 'detector %s AMP %s' % (detNum, det_object)
+            kernels[det_object] = self.generateKernel(xcorrs[det_object], means[det_object], objId)
+        # writing the level might be useful, but might also mess things up xxx
+        # kernels['level'] = self.config.level
+        dataRef.put(kernels)
+
+        self.log.info('Finished generating kernel(s) for %s' % detNum)
+        return pipeBase.Struct(exitStatus=0)
+
+    def _prepareImage(self, exp, gains, level):
+        """Prepare images for cross-correlation calculation.
+
+        For each amp, crop by the border amount, specified by
+        config.nPixBorderXCorr, then rescale by the gain
+        and subtract the sigma-clipped mean.
+        If the level is 'DETECTOR' then this is done
+        for the whole image in-place so that it can be cross-correlated.
+        If the level is 'AMP' then this is done per-amplifier,
+        and each amp-image returned.
```

Comment:
100% agree. I don't know why it said this. I think it's because these docs were adapted from the original code a little too blindly, this was never the case, and certainly code should never work like that!